USWDS - Web components: Beta 2 banner additions

mejiaj opened the pull request:
# Summary

Improvements to Banner. Added JSDocs and component is now configurable via CSS custom properties.

## Related issue

Closes #39. 

## Preview link

Preview link:

[Banner preview →](https://federalist-ab6c0bdb-eccd-4b26-bb5f-b0154661e999.sites.pages.cloud.gov/preview/uswds/web-components/jm/beta-banner-component/iframe.html?args=&id=components-banner--default&viewMode=story) 

## Problem statement

Banner was missing:
- a11y checklist requirements
- theme settings
- JSDoc comments
- Unit tests 

Additionally, the HTTPS lock icon was using unsafe SVG.


## Solution

Added all the items above and swapped the SVG icon for a background image.

## Major changes

### Global CSS properties

Shared global properties created in a new `core` directory. They contain global token values and theme settings. They're meant to be attached at `:root` level and leaves `:host` for component-level settings.

```
.
├── components
└── core
    ├── colors.css
    ├── fonts.css
    └── index.css
```

Core stylesheets also added as an export in `package.json`.

```json
  "exports": {
    "./core/*": "./src/core/*"
  },
```


**Example**

```css
:root {
  // Atom/molecule
  --font-stack-source-sans-pro: "Source Sans Pro", "Helvetica Neue", "Helvetica",
    "Roboto", "Arial", sans-serif;
  --font-stack-public-sans: "Public Sans Web", -apple-system, BlinkMacSystemFont,
    "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji",
    "Segoe UI Emoji", "Segoe UI Symbol";

  // System token
  --usa-font-sans: var(--font-stack-source-sans-pro);
  
  // Theme token
  --usa-font-ui: var(--usa-font-sans);
}
```

### Component properties

> [!CAUTION]
> Missing the theme setting `theme-banner-max-width` because media queries don't allow custom props as a value. Potential solutions: SASS, PostCSS custom media queries, or [Lightning CSS](https://lightningcss.dev/transpilation.html#:~:text=MediaQueries%20%E2%80%93%20shorthand%20for%20MediaIntervalSyntax%20%7C%20MediaRangeSyntax%20%7C%20CustomMediaQueries) each requires additional dependencies for everyone.

| Property                        | Default value              | Fallback value          | Description                   |
| :------------------------------ | :------------------------- | :---------------------- | :---------------------------- |
| --theme-banner-background-color | `--usa-base-lightest`      | `#f0f0f0`               | Sets banner background color. |
| --theme-banner-font-family      | `--usa-font-ui`            | `system-ui, sans-serif` | Sets banner font family.      |
| --theme-banner-link-color       | `--theme-link-color`       | `#005ea2`               | Sets the default link color.  |
| --theme-banner-link-hover-color | `--theme-link-hover-color` | `#1a4480`               | Sets the default link color.  |

Link settings and names match those defined in [USA Link component](https://github.com/uswds/web-components/blob/c6c9553aacb1d42a51b806880f7613a9c3276b98/src/components/usa-link/usa-link.css.js#L5-L8).



### Lock icon replaced

The icon is added via CSS on a class. This allows users to not worry about pasting in a big block of SVG markup. Especially useful when adding custom content via slots.


| Before | After |
| :----- | :---- |
|![image](https://github.com/user-attachments/assets/21b5e08f-990d-4700-923b-74907ae0a09a)|![image](https://github.com/user-attachments/assets/d6826879-7a02-4567-8f6c-b6698a56865e)| 


Following pattern set by @aduth in uswds/uswds#5829.

## Testing and review

1. Visit [Banner preview](https://federalist-ab6c0bdb-eccd-4b26-bb5f-b0154661e999.sites.pages.cloud.gov/preview/uswds/web-components/jm/beta-banner-component/iframe.html?args=&id=components-banner--default&viewMode=story)
1.  Verify there aren't regressions in functionality.
2. Unit tests pass.
3. Theme settings can be overridden.

## Dependency updates

| Dependency name                | Previous version | New version |
| :------------------------------ | :--------------: | :---------: |
| @chromatic-com/storybook       |      1.6.1       |    1.8.0    |
| @rollup/rollup-linux-x64-gnu   |      4.20.0      |   4.21.2    |
| @storybook/addon-a11y          |      8.2.7       |    8.2.9    |
| @storybook/addon-essentials    |      8.2.7       |    8.2.9    |
| @storybook/addon-links         |      8.2.7       |    8.2.9    |
| @storybook/blocks              |      8.2.7       |    8.2.9    |
| @storybook/manager-api         |      8.2.7       |    8.2.9    |
| @storybook/test                |      8.2.7       |    8.2.9    |
| @storybook/theming             |      8.2.7       |    8.2.9    |
| @storybook/web-components      |      8.2.7       |    8.2.9    |
| @storybook/web-components-vite |      8.2.7       |    8.2.9    |
| @uswds/uswds                   |      3.8.1       |    3.8.2    |
| axe-playwright                 |      2.0.1       |    ~2.0.2~ 2.0.3   |
| eslint                         |      9.8.0       |    9.9.1    |
| shadow-dom-testing-library     |      1.11.2      |   1.11.3    |
| storybook                      |      8.2.7       |    8.2.9    |
| vite                           |      5.3.5       |    ~5.4.3~ 5.4.8   |
| **@vitest/eslint-plugin**  |      -       |    1.1.4    |

<!--
For PRs that include dependency updates, uncomment this section and
include a list of the changed dependencies and version numbers.
-->

Commit:
Create global stylesheets for shared CSS vars

Commit:
Create and document Banner settings

Commit:
Move lock icon from markup to CSS

Following pattern from uswds/uswds#5829.

Commit:
Reflect lang attribute

Commit:
Remove redundant story

Commit:
Update dependencies

Commit:
Remove test exclamation

Commit:
Remove commented code

Commit:
Fix hover color

Commit:
Add quotes to background images

Commit:
Add core stylesheets to exports

Commit:
Add colors and minimum story requirements to docs

aduth commented on src/components/usa-banner/index.js:
```diff
@@ -96,32 +121,6 @@ export class UsaBanner extends LitElement {
     return bannerActionText?.textContent;
   }
 
-  // TODO: Use inline image instead or translate strings.
```
Re: "translate strings", would it still be reasonable to expect a text alternative for the image? I think an argument _could_ be made that the image is decorative since it's ["illustrative of adjacent text"](https://www.w3.org/WAI/tutorials/images/decorative/#:~:text=Illustrative%20of%20adjacent%20text%20but%20not%20contributing%20information%20(%E2%80%9Ceye%2Dcandy%E2%80%9D)%3B) ("a lock"), but I'm also not sure how a screen reader would handle effectively empty parentheses `()`.

Maybe either:

1. The `<span />` should be `role=img` with `aria-label`, as it was in https://github.com/uswds/uswds/pull/5829
2. Or the entire parenthesized content should be in a `aria-hidden` wrapper (`<span aria-hidden>(<span class="usa-banner__icon-lock"></span>)</span>`), unless testing confirms that assistive technology gracefully handles the parantheses

mejiaj replied:
@aduth good point. 

I'd consider it decorative, but will manually test in voiceover and find the closest approach to current Banner.

---

### Updated 09/10/24

In stable banner voiceover reads:
> _A lock left parenthesis padlock icon, image or https:// means you've…_

mejiaj replied:
Resolved in ca7978305e1274acccf91541841cc824e11ae9cb.

Commit:
Add eslint plugin for vitest

Following plugin guidance[^1] to avoid ESLint errors in unit tests

[^1]: https://github.com/vitest-dev/eslint-plugin-vitest

Commit:
Separate lock icon to improve HTTPS text readability

Commit:
Improve custom content story

Improve defaults, disable unused controls, and improve content.

Commit:
Create unit tests

Commit:
Update deps; add patch updates to axe playwright and vite

Commit:
Merge branch 'develop' of github.com:uswds/web-components into jm/beta-banner-component

Commit:
Regen custom elements manifest

Commit:
Add font declarations to font stylesheet

Commit:
Run custom element manifest task in parallel with start command

Commit:
Fix name `TLD` comment

mejiaj commented on package.json:
```diff
@@ -57,8 +58,9 @@
     "@storybook/theming": "^8.2.9",
     "@storybook/web-components": "^8.2.9",
     "@storybook/web-components-vite": "^8.2.9",
+    "@vitest/eslint-plugin": "^1.1.4",
```
Helps get rid of errors when writing unit tests.

| Before | After |
| :----- | :---- |
|  ![image](https://github.com/user-attachments/assets/2a5a5485-13af-4ed3-8833-f49667c1325a) |   ![image](https://github.com/user-attachments/assets/daf665e1-3d34-47d5-b260-71a78f9ae542)   |

mahoneycm replied:
Looked into this while doing some POAM for this repo. Good addition 👍

mejiaj commented on src/components/usa-banner/banner.stories.js:
```diff
@@ -4,58 +4,59 @@ import { html, nothing } from "lit";
 
 import { userEvent, expect, waitFor } from "@storybook/test";
 import { within } from "shadow-dom-testing-library";
+import { unsafeHTML } from "lit/directives/unsafe-html.js";
```
> [!TIP]
> Helps render custom slotted content in HTTPS body because it includes HTML.

heymatthenry replied:
**thought:** Ideally we'll be able to do something else here at some point 🤔

mejiaj commented on src/core/colors.css:
> [!TIP]
> Will look into [`@property`](https://developer.mozilla.org/en-US/docs/Web/CSS/@property) now that it has baseline support.

Commit:
Remove duplicate fallback properties

mejiaj commented on storybook/welcome.mdx:
```diff
@@ -29,64 +29,37 @@ npm run start
 <ColorPalette>
   <ColorItem
     title="Base"
     subtitle="Grayscale colors"
     colors={{
-      White: "#FFFFFF",
-      "Longer custom text": "#F3F3F3",
-      Black: "#000",
+      "--usa-base-lightest": "#F0F0F0",
+    }}
+  />
+</ColorPalette>
+
+## Global
+
+<ColorPalette>
+  <ColorItem
+    title="Link"
+    subtitle="Link colors"
+    colors={{
+      "--usa-theme-link-color": "#005EA2",
+      "--usa-theme-link-hover-color": "#1A4480",
```
> [!TIP]
> Starting to document the shared properties here.

mejiaj commented on storybook/welcome.mdx:
```diff
@@ -29,64 +29,37 @@ npm run start
+
+- Variants
+- Properties
+- CSS Custom Properties
```
> [!TIP]
> We'll need to remove this if we continue with custom elements manifest.

mejiaj commented on src/components/usa-banner/index.js:
```diff
@@ -8,9 +8,33 @@ import usFlagSmall from "@uswds/uswds/img/us_flag_small.png";
 import iconHttps from "@uswds/uswds/img/icon-https.svg";
 
+/**
+ * @summary The usa-banner component.
+ *
+ * @attribute {string} lang - The element's language.
+ * @attribute {string} label - The custom aria label users can override.
+ * @attribute {string} tld - The top level domain for the site.
+ *
+ * @cssprop --theme-banner-background-color - Sets banner background color.
+ * @cssprop --theme-banner-font-family - Sets banner font family.
+ * @cssprop --theme-banner-link-color - Sets the default link color.
+ * @cssprop --theme-banner-link-hover-color - Sets the default link color.
+ *
+ * @slot banner-text - The text for official government website text.
+ * @slot banner-action - Action text label "Here's how you know."
+ * @slot domain-heading - Heading text for the domain section.
+ * @slot domain-text - Body text for domain section.
+ * @slot https-heading - Heading for HTTPs section.
+ * @slot https-text - Body text for HTTPs section.
+ *
+ * @tagname usa-banner
```
> [!note]
> StorybookJS docs are generated from this. There's currently an issue for hiding internal props. A small downside, but still benefits us in not having to re-write these in Story Docs.

**Example**
![banner-docs](https://github.com/user-attachments/assets/74f5df0e-d1c6-48bc-9c5d-b918541ca984)

Commit:
Merge branch 'develop' of github.com:uswds/web-components into jm/beta-banner-component

Commit:
Merge branch 'develop' of github.com:uswds/web-components into jm/beta-banner-component

Commit:
Resolve conflicts and regen generated files

Commit:
Add a script to build and watch for changes

Useful for testing in other environments without having to run build on every change.

mejiaj commented on package.json:
```diff
@@ -17,18 +17,19 @@
-    "build": "npm run clean && vite build && npm run manifest:build",
-    "clean": "node ./scripts/rmrf.js dist",
+    "build": "vite build && npm run manifest:build",
+    "build:watch": "vite build --watch",
```
> [!TIP]
> Added for local development and testing. For example, when using `npm link` to test in Sandbox environment. You can run this without having to manually build every update.

heymatthenry replied:
**praise:** Nice. Thanks. I'd been doing this by hand a lot when `npm link`ing

mejiaj commented:
@heymatthenry I've done the the following:

- Created a new blank vite environment [[vite-sandbox](https://github.com/uswds/uswds-sandbox/tree/vite-sandbox)] on uswds-sandbox.
- Created a branch for testing the `currentDirective._$initialize is not a function` issue, that's [[`vite-sandbox-test-elements`](https://github.com/uswds/uswds-sandbox/tree/vite-sandbox-test-elements)].
- Opened a draft PR with findings/workarounds, [USWDS-Sandbox - Test: Banner Elements](https://github.com/uswds/uswds-sandbox/pull/170). 
  - Can you test and confirm that this is working?

Commit:
Merge branch 'develop' of github.com:uswds/web-components into jm/beta-banner-component

Commit:
Remove global styles that were shrinking storybook previews

Commit:
Regenerate custom elements manifest

mejiaj commented on storybook/index.css:
```diff
@@ -68,14 +24,6 @@
-  min-width: 320px;
-  min-height: 100vh;
-}
-
```
> [!NOTE]
> Removed these styles because it was affecting our storybook previews. Screenshot below.

![capture -Arc-2025-01-08@2x](https://github.com/user-attachments/assets/31e8c516-0ce4-4acf-9174-fadd9cab4662)

mejiaj commented:
Ready for re-review. Resolved conflicts and added a [comment](https://github.com/uswds/web-components/pull/65#discussion_r1907905036) on recent style change in da0454a

heymatthenry commented on src/core/fonts.css:
```diff
@@ -0,0 +1,211 @@
+  font-weight: 700;
+  src: url("@uswds/uswds/fonts/public-sans/PublicSans-BoldItalic.woff2")
+    format("woff2");
+}
```
**thought:** These will all need to be in their own package sooner rather than later (but not today)

heymatthenry reviewed (approved):
Great work on this! Especially on de-Sass-ifying things!
